WalletWasabi/Wallets/UnconfirmedTransactionChainProvider.cs

```diff
@@ -0,0 +1,133 @@
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class UnconfirmedTransactionChainProvider : BackgroundService
+{
+	private const int MaximumDelayInSeconds = 120;
+	private const int MaximumRequestsInParallel = 3;
+
+	public UnconfirmedTransactionChainProvider(WasabiHttpClientFactory httpClientFactory)
+	{
+		HttpClient = httpClientFactory.NewHttpClient(httpClientFactory.BackendUriGetter, Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+	}
+
+	public event EventHandler<EventArgs>? RequestedUnconfirmedChainArrived;
+
+	public ConcurrentDictionary<uint256, List<UnconfirmedTransactionChainItem>> UnconfirmedChainCache { get; } = new();
+
+	private IHttpClient HttpClient { get; }
+
+	private Channel<uint256> Channel { get; } = System.Threading.Channels.Channel.CreateUnbounded<uint256>();
+
+	private async Task FetchUnconfirmedTransactionChainAsync(uint256 txid, CancellationToken cancellationToken)
+	{
+		const int MaxAttempts = 3;
+
+		for (int i = 0; i < MaxAttempts; i++)
+		{
+			try
+			{
+				using CancellationTokenSource timeoutCts = new(TimeSpan.FromSeconds(60));
+				using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
+
+				var response = await HttpClient.SendAsync(
+					HttpMethod.Get,
+					$"api/v{Helpers.Constants.BackendMajorVersion}/btc/Blockchain/unconfirmed-transaction-chain?transactionId={txid}",
+					null,
+					linkedCts.Token).ConfigureAwait(false);
+
+				if (response.StatusCode != HttpStatusCode.OK)
+				{
+					await response.ThrowRequestExceptionFromContentAsync(cancellationToken).ConfigureAwait(false);
+				}
+
+				var unconfirmedChain = await response.Content.ReadAsJsonAsync<UnconfirmedTransactionChainItem[]>().ConfigureAwait(false);
+
+				if (!UnconfirmedChainCache.TryAdd(txid, unconfirmedChain.ToList()))
+				{
+					throw new InvalidOperationException($"Failed to cache unconfirmed tx chain for {txid}");
+				}
+
+				RequestedUnconfirmedChainArrived?.Invoke(this, EventArgs.Empty);
+
+				return;
+			}
+			catch (OperationCanceledException)
+			{
+				Logger.LogTrace("Request was cancelled by exiting the app.");
+			}
+			catch (Exception ex)
+			{
+				Logger.LogWarning($"Attempt: {i}. Failed to fetch transaction fee. {ex}");
+			}
+		}
+	}
+
+	public void CheckAndScheduleRequestIfNeeded(SmartTransaction tx)
+	{
+		if (!tx.Confirmed && tx.ForeignInputs.Count != 0 && !UnconfirmedChainCache.ContainsKey(tx.GetHash()))
+		{
+			Channel.Writer.TryWrite(tx.GetHash());
+		}
+	}
+
+	protected override async Task ExecuteAsync(CancellationToken cancel)
+	{
+		List<Task> tasks = [];
+		while (!cancel.IsCancellationRequested)
+		{
+			var txidToFetch = await Channel.Reader.ReadAsync(cancel).ConfigureAwait(false);
+
+			tasks.Add(Task.Run(() => ScheduledTask(txidToFetch), cancel));
+
+			if (tasks.Count >= MaximumRequestsInParallel)
+			{
+				Task completedTask = await Task.WhenAny(tasks).ConfigureAwait(false);
```

Comment:
You need to await `completedTask`. Be aware: it can throw.

Reply:
Why? 

`completedTask` has no return value and `ScheduledTask` catches and logs every exception.

Reply:
It is best practice.